src/main/kotlin/org/rust/ide/annotator/CargoCheckExtAnnotator.kt

```diff
@@ -0,0 +1,630 @@
+                )
+
+            }
+
+        }
+
+        //file - The file to annotate
+        fun doAnnotate(holder: AnnotationHolder)  {
+            message.doAnnotate(holder, target)
+        }
+    }
+
+
+}
+
+data class TEAAnnotationInfo(val file: PsiFile, val editor: Editor)
+
+class TEAAnnotationResult(commandOutput: String) {
+
+    var parsedMessages = emptyList<CargoJson.TopMessage>()
+
+    private companion object {
+        val jsonParser : Gson =
+            GsonBuilder()
+                .registerTypeAdapter(CargoJson.TopMessage.Message::class.java,
+                    CargoJson.TopMessage.Message.Companion.MessageFactory())
+                .create()
+
+    }
+
+    init {
+
+        parsedMessages += commandOutput.lines().filter {
+            //Apparently some lines also have human readable messages. We ignore any lines not beginning with '{'
+            val trimmed = it.trimStart()
+            trimmed.isNotEmpty() && trimmed[0] =='{'
+        }.filter {
+            //We check the first field name by finding the first " then reading until another " (we expect
+            //that a '{' present and don't check for it).
+            val first = it.indexOfFirst { it == '"' }
+            assert(first != -1)
+            val second = it.indexOf('"', first + 1)
+            assert(second != -1)
+
+            when (it.subSequence(first + 1, second)) {
+                "message" -> true
+                "features" -> false
+                else -> {assert(false); throw AssertionError()}
+            }
+
+        } .map {
+            //Next parse the messages
+            //We expect one message per line
+            jsonParser.fromJson(it, CargoJson.TopMessage::class.java)
+        }
+    }
+
+    fun prettyJson(): String {
+        return "${jsonParser.toJson(parsedMessages)}\n"
+    }
+
+    override fun toString(): String {
+        return prettyJson()
+    }
+}
+
+class CargoCheckExtAnnotator: ExternalAnnotator<TEAAnnotationInfo, TEAAnnotationResult>() {
+
+    //We use this server to cache the cargo check result instead of running it for every file
+    object CargoCheckServer {
+
+        //The cache map (project -> last result)
+        private val cache = ConcurrentHashMap<Project, TEAAnnotationResult>()
+
+        private fun doCargoCheck(project : Project) : TEAAnnotationResult {
+            IJDebug.log("beginning cargo check run for project ${project.name}")
+
+            val rustcExec : String = "/home/developer/.cargo/bin/cargo"
+            val path = project.basePath
+
+            val args = listOf(rustcExec, "check", "--manifest-path=$path/Cargo.toml", "--message-format=json")
+            val cmd = args.joinToString(" ")
+
+            IJDebug.log("executing command `$cmd`")
+
+            val err : String
+            val out : String
+            val retVal : Int
+            var process : Process? = null
+
+            try {
+
+                process = ProcessBuilder(args).start()
```

Comment:
Can we use `CapturingProcessHandler` instead? It already can split output by lines. 

Or perhaps we can just synchronously wait for all the output to be available?